main.cpp

```diff
@@ -1,48 +1,168 @@
+#include <algorithm>
+#include <stack>
+#include <map>
 using namespace std;
 
 struct node {
     int key;
     node *left;
     node *right;
+	node *par;
 };
 
-void insert(int key, node *root, node *last) {
-    
+node *buildtree(int key, node *left, node *right) {
+    //builds tree in bottom up way
+	
+    node *root = new node;
+    root -> key = key;
+    root -> left = left;
+    root -> right = right;
+	root -> par = NULL;
+	if (root -> left != NULL) {
+		root -> left -> par = root;
+	}
+	if (root -> right != NULL) {
+		root -> right -> par = root;
+	}
+	
+    return root;
+}
+
+void printtree(node *root) {
+	//for debugging purpose
+	
+    cout<< root -> key << endl;
+    if (root -> left != NULL) {
+		printtree(root -> left);
+    }
+	if (root -> right != NULL) {
+		printtree(root -> right);
+	}
 }
 
-bool found(int key, node *root) {
-    if (root->left == NULL && root->right == NULL) 
-        return key == root->key;
-    else if (root -> left == NULL) 
-        return found(key, root->right);
-    else if (root -> right == NULL)
-        return found(key, root->left);
-    else 
-        return found(key, root->left) || found(key, root->right);
+void printancestors(vector<int> anc) {
+	if (anc.empty()) {
+		cout << "No ancestors found";
+	}
+	
+    for (int i = 0; i < anc.size(); i++) {
+        cout << anc[i] << " ";
+    }
+	cout << endl;
 }
 
-void printancestors(int key, node *root) {
-    if (found(key, root->left)) {
-        printf("%d ", root->key);
-        printancestors(key, root->left);
-    } else if (found(key, root->right)) {
-        printf("%d ", root->key);
-        printancestors(key, root->right);
-    } else 
-        throw "Key doesn't exist in this tree";
+node *dfs(int key, node *root) {
+	/*returns the node that has key we are looking for
+	and finds parents
+	 returns NULL if the key doesn't exist. */
+	
+	stack<node*> st;
+	st.push(root);
+	
+	node *curr = root;
+	
+	bool found = false;
+	
+	while (!st.empty() && !found) {
+		curr = st.top();
+		st.pop();
+		
+		if (curr -> key == key) {
+			found = true;
+		}
+		
+		if (curr -> left != NULL) {
+			st.push(curr -> left);
+		}
+		
+		if (curr -> right != NULL) {
+			st.push(curr -> right);
+		}
+	}
+	
+	if (!found) {
+		return NULL;
+	}
+	
+	return curr;
 }
 
-void lowestcommonancestor(int key1, int key2, node *root) {
+vector<int> findancestors(int key, node *root) {
+    vector<int> ancestors;
     
+    node *curr = dfs(key, root);
+    
+	while (curr -> par != NULL) {
+		ancestors.push_back(curr -> par -> key);
+		curr = curr -> par;
+	}
+	
+    return ancestors;
 }
 
+int lowestcommonancestor(int key1, int key2, node *root) {
+    vector<int> key1ancestors = findancestors(key1, root);
+    vector<int> key2ancestors = findancestors(key2, root);
+	sort(key1ancestors.begin(), key1ancestors.end());
+	sort(key2ancestors.begin(), key2ancestors.end());
+	
+	
+	int i = 0;
+	int j = 0;
+	
+	while (i < key1ancestors.size() && j < key2ancestors.size()) {
```

Comment:
If I am correct, now the code has a different bug: It returns the common ancestor with the highest key, which is not necessarily the lowest (in terms of distance from the tree root) ancestor.
For example, for following tree:
```
  3
 /
2
 \
  1
 / \
4   5
```
When asking for the lowest common ancestor of `4` and `5`, sorted ancestors vectors are `{1,2,3,4}` and `{1,2,3,5}` accordingly. Your solution will return `3`, while the lowest common ancestor is `1`.

Suggested method to fix this:
1. First, write a test function (similar to `testbalancedtree`) with an input that makes this bug return an incorrect output. Run the test to see that it happens.
2. Fix your code
3. Run the test again to see that the bug was fixed.